### PR TITLE
Relaxes privilege test for macvlan creation

### PR DESCRIPTION
**Pull request by cjoster**
The preemptive test for any privileged container before creating a macvlan interface in response to a request for `pod.network.openshift.io/assign-macvlan` annotation is superficial and incorrect.

While privileges are required to manipulate network configuration, full privilege is an overset. CAP_NET_ADMIN (along with uid == 0) is all that is required to manipulate the relevant configuration regarding a macvlan interface.

The original code does not make any attempt to determine if the container that will be doing the configuration is the privileged container--how could it--nor does this addition.  This is merely a relaxation that allows the pod to start in the presence of either a privileged container, or an unprivilged container that but with CAP_NET_ADMIN--at least requested, if possibly not granted.

### CODE REVIEW

**Comment by danwinship** [comment]
There should probably be a comment explaining this, but the overprotective check is intentional.

The problem is that having CAP_NET_ADMIN plus a macvlan attached to the node's ethernet gives you more power than is immediately obvious. In particular, it gives you the power to snoop and spoof VXLAN traffic between nodes, and to DoS nodes by spoofing ARP messages from them or from the router. And maybe other things. I never bothered to figure out the full set of possibilities. Basically though, you don't want to let a pod have CAP_NET_ADMIN plus a macvlan unless you absolutely 100% trust it, and the easiest way to express that was to force the admin to make the pod be privileged.

File: pkg/network/node/pod.go [hunk 1]
```diff
@@ -347,15 +347,29 @@ func maybeAddMacvlan(pod *kapi.Pod, netns string) error {
...
+	netprivs := false
 	for _, container := range append(pod.Spec.Containers, pod.Spec.InitContainers...) {
-		if container.SecurityContext != nil && container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged {
-			privileged = true
+		if container.SecurityContext != nil {
+			if container.SecurityContext.Privileged != nil && *container.SecurityContext.Privileged {
+				netprivs = true
+				break
+			}
+			if !netprivs && container.SecurityContext.Capabilities != nil && container.SecurityContext.Capabilities.Add != nil &&
```
Comment (rajatchopra) [inline pkg/network/node/pod.go]: netprivs will always be false here, so no need to check it.

**Comment by cjoster** [comment]
@danwinship ...

1) If this is the intent of the existing code, it should test for privileged *AND* NET_ADMIN, which it doesn't--admittedly neither does my change. A cluster admin could well create a "privileged" SCC that does not have NET_ADMIN and the existing check would pass, but the container would fail to setup its networking.

2) While NET_ADMIN is a lofty privilege set, its nonetheless required to manipulate networking parameters.

3) In a world where, "every privilege that is not required should be stripped," telling a user "you need to assign 100x additional privilege in order to get the ones you want," is entirely incongruent with that.

4) Only with NET_RAW can frames/packets be spoofed, which, admittedly is already in the bounding set. I'd be open to a NET_ADMIN without NET_RAW option, but it's worth mentioning that both are required to manipulate iptables (probably considerably more).

5) A user may well want a macvlan interface without NET_ADMIN. I can't think of a good reason to do that, but there's no reason to stop a user from doing so. NET_RAW concerns aside.

6) Snooping on traffic would require the macvlan0 interface to have the same MAC address as the host node, which the kernel prevents. Only broadcast frames will be copied to the container, the security or lack thereof is a know issue in broadcast frames, but in any case, is neither the concern nor job of higher layers.

7) Any other layer-2 attack, which again, requires NET_RAW, can be done to the vxlan network by a pod anyway. Granted, that only affects other pods, but the original code also does not prevent this.

In conclusion, gating one privilege by the requirement of assigning hundreds of others can't be called "secure" any more than gaining the single privilege and encourages carelessly handing out privileges. It's worth having a discussion regarding NET_RAW, and possibly the host setting up the macvlan interface, but telling a user that they a huge superset of permissions is required to get just one--in an effort to protect users from themselves--without even checking for the correct capability in the first place is incorrect.

**Comment by knobunc** [comment]
Given that you can install iptables rules to intercept traffic for the entire node if you have CAP_NET_ADMIN it feels to me like it's a serious problem.  We definitely need to document why the code is enforcing that they are privileged, and it looks like we need to add a check for CAP_NET_ADMIN too... but I really don't want to make this too easy to do.  We could consider splitting out a finer grained permission, but I'm not sure if it's worth the complexity.

@simo5 Any opinions?  Thanks

**Comment by danwinship** [comment]
> If this is the intent of the existing code, it should test for privileged AND NET_ADMIN, which it doesn't--admittedly neither does my change. A cluster admin could well create a "privileged" SCC that does not have NET_ADMIN and the existing check would pass, but the container would fail to setup its networking.

It's not checking the SCC, it's checking the pod spec. A privileged pod effectively has all capabilities.

> Any other layer-2 attack, which again, requires NET_RAW, can be done to the vxlan network by a pod anyway.

The SDN's OVS flows require all packets coming out of a pod to have the correct IP and MAC address.

**Comment by simo5** [comment]
Is there a use case for allowing pods to use macvlans without being privileged ?

**Comment by danwinship** [comment]
> Given that you can install iptables rules to intercept traffic for the entire node if you have CAP_NET_ADMIN

No, because you would have CAP_NET_ADMIN *inside the pod's namespace*. You can't do anything to traffic in the main namespace.

> it looks like we need to add a check for CAP_NET_ADMIN

If we're checking for privileged, we don't need to check for CAP_NET_ADMIN.

> We could consider splitting out a finer grained permission, but I'm not sure if it's worth the complexity.

It's definitely true that it would be good to not force admins to give the container tons of privileges that it doesn't need. But unless someone can convince me that CAP_NET_ADMIN plus a macvlan isn't dangerous, then I want to make it *look* dangerous.

**Comment by danwinship** [comment]
> Is there a use case for allowing pods to use macvlans without being privileged ?

That's not really the question. The feature (the `pod.network.openshift.io/assign-macvlan` annotation) really only exists to support the egress router, and the egress router pod needs to create iptables rules inside the pod network namespace. So there's no use case for having the macvlan without having at least CAP_NET_ADMIN.

The question is whether it's safe to give untrusted users CAP_NET_ADMIN in a namespace that contains a macvlan attached to the node's primary ethernet interface.

**Comment by danwinship** [comment]
> Snooping on traffic would require the macvlan0 interface to have the same MAC address as the host node, which the kernel prevents.

It doesn't prevent it from claiming the IP/MAC of a *different* node on the same subnet though

**Comment by knobunc** [comment]
I think C_N_A + a macvlan IS dangerous.  I'd be happy if we just added the rationale as a comment to the code and closed out the bug.

**Comment by danwinship** [comment]
And people are working on solutions upstream for "privileged sidecars" in the context of Istio, which should either make it possible for the egress-router to fiddle with iptables rules without needing CAP_NET_ADMIN, or else to better specify the security concerns associated with granting the privilieges. Or both.

**Comment by cjoster** [comment]
> If we're checking for privileged, we don't need to check for CAP_NET_ADMIN.

Couldn't you, with require_drop, concievable make a "privileged" scc that doesn't have CAP_NET_ADMIN?

> The question is whether it's safe to give untrusted users CAP_NET_ADMIN in a namespace that contains a macvlan attached to the node's primary ethernet interface.

At least for the immediate need, I totally agree that a user using an scc that hands out CAP_NET_ADMIN with a macvlan interface should still be a trusted user. I'm suggesting that it's perfectly OK to trust the user, but not trust the software running in the container to be free of exploit or otherwise abuse the additional capabilities that granting full privilege to the container would amount.

**Comment by danwinship** [comment]
More discussion on https://trello.com/c/ZGfEFWf3/75-allow-non-privileged-egress-router-egresssecurity

The long-term solution is to use the "multiple networks" support being worked on upstream; the egress-router would an extra CNI plugin to set up the macvlan, and access to that plugin would be controlled by the same security mechanisms used by the multi-network feature in general.

If we want to fix it before then, the fix is probably to make the SecurityContextConstraint admission controller treat the `assign-macvlan` annotation as equivalent to `hostNetwork: true` (ie, only allowing a user to create/edit a pod with that annotation if they have access to an SCC with `allowHostNetwork: true`), and drop the requirement that egress-routers be privileged. There are possibly some security issues with migrating from the current solution to the proposed solution that might require we do some checks/fixing during a cluster upgrade.
